src/celeritas/UnitTypes.hh

```diff
@@ -0,0 +1,298 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/UnitTypes.hh
+//! \brief Annotated unit struct definitions for use with Quantity
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "celeritas_config.h"
+#include "corecel/math/UnitUtils.hh"
+
+#include "Constants.hh"
+#include "Units.hh"
+
+namespace celeritas
+{
+namespace units
+{
+//---------------------------------------------------------------------------//
+//!@{
+//! \name Natural units
+
+struct CLight
+{
+    static CELER_CONSTEXPR_FUNCTION real_type value()
+    {
+        return constants::c_light;
+    }
+    static char const* label() { return "c"; }
+};
+
+//! "Natural units" for energy
+struct Mev
+{
+    static CELER_CONSTEXPR_FUNCTION real_type value()
+    {
+#if CELERITAS_UNITS == CELERITAS_UNITS_CLHEP
+        return units::megaelectronvolt;
+#else
+        return real_type(1e6) * constants::e_electron * units::volt;
+#endif
+    }
+    static char const* label() { return "MeV"; }
+};
+
+//! "Natural units" for mass
+struct MevPerCsq : UnitDivide<Mev, UnitProduct<CLight, CLight>>
+{
+    static char const* label() { return "MeV/c^2"; }
+};
+
+//! "Natural units" for momentum
+struct MevPerC : UnitDivide<Mev, CLight>
+{
+    static char const* label() { return "MeV/c"; }
+};
+
+//! Magnitude of electron charge (positive sign)
+struct EElectron
+{
+    static CELER_CONSTEXPR_FUNCTION real_type value()
+    {
+#if CELERITAS_UNITS == CELERITAS_UNITS_CLHEP
+        return units::e_electron;  // Mathematically equal to constants
+#else
+        return constants::e_electron;
```

Comment:
(starting a new conversation to focus on a single point, see https://github.com/celeritas-project/celeritas/pull/1076/files#r1439657682)

> The units::e_electron here should be mathematically equal to constants::e_electron, but might not be numerically.

I think this is wrong and opens up to subtle inconsistencies.  In a unit system where `e_electron` is 1 there is 3 ways to get that value (and 2 other-wise)
* `units::e_electron`
* `constants::e_electron`
* `EElectron::value`
but if the above is true there is no way to guarantees the value are the same and unless we have an automated way to enforce only one is used (I guess given its existence it would be `EElectron::value`, we will eventually have code that use directly or indirectly all 3.

In other in a unit system that has decided that `e_electron` is 1, I would expect that `constants::e_electron` is that value exactly (since the physics constant in `constants` are expressed in the native unit system.

Reply:
Fair enough. I've made a conditional expression in `Constants.hh` to guarantee both are exactly unity in CLHEP.